### PR TITLE
Remove onRequest function and add onGet, onPost functions

### LINKED ISSUE #145: Implement 'onPost', 'onGet' functions; deprecate 'onRequest' function
Implement `onPost`, `onGet` server-side JS functions instead of the `onRequest` method.

### PR DESCRIPTION
**Pull request by Shan1024**
Resolves #145, #125

### CODE REVIEW

File: components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java [hunk 1]
```diff
@@ -83,6 +90,18 @@ protected void compile(String scriptSource) {
...
+
+        Set<String> keySet = ((ScriptObjectMirror) engineBindings.get("nashorn.global")).keySet();
+        if (keySet.contains(FUNCTION_ON_GET)) {
+            isOnGetFound = true;
+        }
+        if (keySet.contains(FUNCTION_ON_POST)) {
+            isOnPostFound = true;
+        }
+        if (!isOnGetFound && !isOnPostFound) {
+            throw new UUFException("'onGet' and 'onPost' functions not found. Please implement at least on of them.");
```
Comment (this) [inline components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java]: `Either 'FUNCTION_ON_GET' nor 'FUNCTION_ON_POST' cannot be found in JavaScript file 'absolutePath'. Please implement at least one of them.`

File: components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java [hunk 2]
```diff
@@ -83,6 +90,18 @@ protected void compile(String scriptSource) {
         } catch (ScriptException e) {
             throw new UUFException("An error occurred while evaluating the JavaScript file '" + absolutePath + "'.", e);
         }
+
+        Set<String> keySet = ((ScriptObjectMirror) engineBindings.get("nashorn.global")).keySet();
```
Comment (this) [inline components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java]: rename this variable to `availableFunctions`

File: components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java [hunk 3]
```diff
@@ -48,11 +50,16 @@
...
 
+    private static final String FUNCTION_ON_GET = "onGet";
+    private static final String FUNCTION_ON_POST = "onPost";
+
     private final NashornScriptEngine engine;
     private final UUFBindings engineBindings;
     private final String absolutePath;
     private final String relativePath;
     private final String componentPath;
+    private boolean isOnGetFound;
```
Comment (this) [inline components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java]: rename this variable to `hasOnGetFunction`; same for the next

File: components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java [hunk 4]
```diff
@@ -104,15 +123,24 @@ protected String getRelativePath() {
...
+            } else {
+                return null;
+            }
         } catch (ScriptException e) {
-            throw new UUFException("An error occurred when executing the 'onRequest' function in JavaScript file '" +
-                                           absolutePath + "' with context '" + context + "'.", e);
+            throw new UUFException(
+                    "An error occurred when executing the '" + functionName + "' function in JavaScript file '" +
+                            absolutePath + "' with context '" + context + "'.", e);
         } catch (NoSuchMethodException e) {
```
Comment (this) [inline components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java]: This won't happen as we check for the existence of `onGet`, `onPost` functions, so we can remove this catch-block.

File: components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java [hunk 5]
```diff
@@ -104,15 +123,24 @@ protected String getRelativePath() {
 
     @Override
     public Object execute(Object context, API api) {
+        String functionName = null;
         try {
             engineBindings.setJSFunctionProvider(new JsFunctionsImpl(api));
-            return engine.invokeFunction("onRequest", context);
+            functionName = api.getRequestLookup().getRequest().isGet() ? FUNCTION_ON_GET : FUNCTION_ON_POST;
```
Comment (this) [inline components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java]: Change the logic as follows
```java
if(isGet())
   return hasOnGetFunction?  <invoke FUNCTION_ON_GET>: null;
else
   return hasOnPostFunction? <invoke FUNCTION_ON_POST>: null;
```

File: components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java (hunk 4 above)
Comment (Shan1024) [inline components/uuf-renderablecreator-hbs/src/main/java/org/wso2/carbon/uuf/renderablecreator/hbs/impl/JsExecutable.java]: We need to catch the `NoSuchMethodException` which is thrown by `engine.invokeFunction()` method. So we need this `catch`.

File: components/uuf-httpconnector-msf4j/src/main/java/org/wso2/carbon/uuf/httpconnector/msf4j/MicroserviceHttpRequest.java [hunk 6]
```diff
@@ -234,6 +238,11 @@ public int getRemotePort() {
     }
 
     @Override
+    public boolean isGet() {
```
Comment (this) [inline components/uuf-httpconnector-msf4j/src/main/java/org/wso2/carbon/uuf/httpconnector/msf4j/MicroserviceHttpRequest.java]: Let's move this method to be after `getMethod()` method. Similar to the `HttpRequest` interface.

Comment (Shan1024) [inline components/uuf-httpconnector-msf4j/src/main/java/org/wso2/carbon/uuf/httpconnector/msf4j/MicroserviceHttpRequest.java]: Method moved.

File: components/uuf-core/src/main/java/org/wso2/carbon/uuf/spi/HttpRequest.java [hunk 7]
```diff
@@ -38,6 +38,12 @@
     String getMethod();
 
     /**
+     * Return whether this request is a GET request or not.
+     * @return {@code true} if this is a GET request and {@code false} if this is a POST request
+     */
+    boolean isGet();
```
Comment (Kishanthan) [inline components/uuf-core/src/main/java/org/wso2/carbon/uuf/spi/HttpRequest.java]: Shall we rename this this "isGetRequest" ?

File: samples/apps/org.wso2.carbon.uuf.sample.features-app.app/src/main/pages/file-upload.js [hunk 8]
```diff
@@ -14,25 +14,27 @@
...
-        var tempDirPath = System.getProperty('java.io.tmpdir');
-        var destination = Paths.get(tempDirPath).resolve(uploadedFile.name);
-        var sourcePath = Paths.get(uploadedFile.path);
-        var destinationPath = Paths.get(destination);
+}
 
-        //copy the selected file to UUF_HOME/tmp directory overriding any existing content with the same name
-        Files.copy(sourcePath, destinationPath, StandardCopyOption.REPLACE_EXISTING);
+function onPost(env) {
+    var Paths = Java.type('java.nio.file.Paths');
```
Comment (Kishanthan) [inline samples/apps/org.wso2.carbon.uuf.sample.features-app.app/src/main/pages/file-upload.js]: We should not use this approach like writing java code with javascript but rather this should be moved to a java class level and then call that calls within this function call. proper practice is to separate the java code from javascript.

Comment (this) [inline samples/apps/org.wso2.carbon.uuf.sample.features-app.app/src/main/pages/file-upload.js]: This will be fixed in PR #155

**Pull request merged**